Generalized flip

erensezener opened the pull request:
This PR generalizes `fliplr` and `flipud` for arbitrary axes. 

`flipud`  and `fliplr` reverse the elements of an array along `axis=0` and `axis=1` respectively. The `flip` function reverses the elements of an array along any given axis. In case `flip` is called with  `axis=0` or  `axis=1`, the function is equivalent to `flipud`  and `fliplr` respectively.

A similar function is also available in MATLAB™.

This pull request is part of a BCCN Berlin student project by @denisalevi @ge00rg @erensezener

erensezener committed:
ENH: Add generalized flip function and its tests

shoyer commented:
I think this is a nice idea, *but* we have a policy that new API needs to be discussed on the mailing list. Could you please post your proposal to the numpy-discussion list: https://mail.scipy.org/mailman/listinfo/numpy-discussion

Thanks!

madphysicist commented on numpy/lib/tests/test_twodim_base.py:
```diff
@@ -144,6 +144,77 @@ def test_failure(self):
+                    [7, 6]]])
+
+        assert_equal(flip(a, 2), b)
+
```
I think you should add a test for a flip at higher dimensions. If you don't want to construct the input array and expected result manually, perhaps show something like

    a = np.arange(2 * 3 * 4 * 5).reshape(2, 3, 4, 5)
    for i in range(a.ndim):
        assert_equal(flip(a, i), np.flipud(a.swapaxes(0, i)).swapaxes(i, 0))

madphysicist replied:
This will also implicitly demonstrate the claim that the shape of the array is preserved, although you can add an explicit check. All your other test arrays are 2**3, so it would be hard to tell if the dimensions moved around.

erensezener replied:
Added your test example here: 468c4f2

madphysicist commented on numpy/lib/twodim_base.py:
```diff
@@ -32,6 +32,76 @@ def _min_int(low, high):
+    """
+    Reverse the order of elements in an array along the given axis.
+
+    All other axes are preserved, but appear in a different order than before.
```
"The shape of the array is preserved, but the elements are reordered."?

shoyer commented on numpy/lib/twodim_base.py:
```diff
@@ -32,6 +32,76 @@ def _min_int(low, high):
+        m = asanyarray(m)
+        if m.ndim < 2:
+            raise ValueError("Input must be >= 2-d.")
+        return (m.swapaxes(axis, 0)[::-1,...]).swapaxes(0, axis)
```
Rather than using `swapaxes` and making all these special cases, I would simply construct a tuple of the appropriate length with the appropriate `slice` objects.

Note that these are equivalent:
```
m[1, 2:3, ::-1, :]

index = (1, slice(2, 3), slice(None, None, -1), slice(None))
m[index]
```

In fact, the former is pretty much syntactic sugar for the later.

mhvk replied:
Nice to generalize, but might as well not have the special cases, as @madphysicist notes.  Do think you want to support negative values of `axis`, though, so, combining with @shoyer's comment, it would just be:
```
m = asanyarray(m)
if axis < 0:
    axis += m.ndim
if axis < 0 or axis >= m.mdim:
    raise ValueError(...)

return m[(slice(None),) * axis + (slice(None, None, -1),)]
```

madphysicist replied:
@mhvk It would be inconsistent with the rest of `numpy` not to support negative axes.

erensezener replied:
> Rather than using ```swapaxes``` and making all these ```special cases```, I would simply construct a tuple of the appropriate length with the appropriate slice objects.

@shoyer isn't the version with  ```swapaxes``` cleaner? 

@denisalevi will test both versions and report the speed difference.

shoyer replied:
There's a lot more going on under the hood (in NumPy) when you use `swapaxes` twice (you construct two intermediate arrays), so in my view using the tuple for indexing more elegant. I suppose it's a matter of taste.

Another thing I like about omitting `swapaxes` is that it makes it easier to support non-numpy ndarray objects (e.g., dask.array), so `np.flip(m, axis=2)` would work on a dask array out of the box.

Towards that end, I would also enhance duck typing support by writing:
```
if not hasattr(m, 'ndim'):
    m = asarray(m)
```
That way, third party libraries could add flip support to their arrays simply by ensuring that indexing works.

njsmith replied:
If you want to get really fancy/simple, you could handle the negative indices at the same time as constructing the indexing object...
```
indexer = [slice(None)] * arr.ndim
indexer[axis] = slice(None, None, -1)
return arr[tuple(indexer)]
```
(taking advantage of the fact that list indexing has the same special handling of negative indices and bounds checking as what we want for axes)

erensezener replied:
@shoyer So something like this instead of using swapaxes?

```python 
return m[tuple(slice(None, None, -1) if i == axis else slice(None) 
                                                          for i in range(m.ndim))]
```

shoyer replied:
@erensezener Yes, that's one of several equally good ways to write it (@njsmith and @mhvk showed other approaches as well).

erensezener replied:
Addressed this issue with abe1735. I used @njsmith's suggestion since it also handles negative indices.

denisalevi replied:
I checked the speed for a few different implementations using ```timeit``` and flipping ```a = np.arange(100**4).reshape(100,100,100,100)```.

Our implementation using ```swapaxes``` is the fastest, but using
<pre><code class="python">def flip(m, axis):
    idx = [slice(None)] * m.ndim
    idx[axis] = slice(None, None, -1)
    return m[idx]
</code></pre>
as suggested by @njsmith is only slighlty slower (1000000 loops, best of 3: 1.3us and 1.4us, respectively).

madphysicist commented on numpy/lib/twodim_base.py:
```diff
@@ -32,6 +32,76 @@ def _min_int(low, high):
+    >>> np.all(flip(A,2)==A[:,:,::-1,...])
+    True
+    """
+    if axis==0:
```
Are these special cases really necessary? The function seems to be able to handle them already.

erensezener replied:
They are not necessary but ```flipud``` and ```fliplr``` should be slightly faster. Should we remove the cases?

njsmith replied:
In general, clarity is more important than unmeasured micro-optimizations.

On Fri, Feb 26, 2016 at 3:10 PM, C. Eren Sezener
<notifications@github.com> wrote:
> In numpy/lib/twodim_base.py:
>
>> +
>> +           [[0, 1],
>> +            [2, 3]]])
>> +
>> +    >>> flip(A, 1)
>> +    array([[[2, 3],
>> +            [0, 1]],
>> +
>> +           [[6, 7],
>> +            [4, 5]]])
>> +
>> +    >>> A = np.random.randn(3,4,5)
>> +    >>> np.all(flip(A,2)==A[:,:,::-1,...])
>> +    True
>> +    """
>> +    if axis==0:
>
> They are not necessary but flipud and fliplr should be slightly faster.
> Should we remove the cases?
>
> —
> Reply to this email directly or view it on GitHub.



-- 
Nathaniel J. Smith -- https://vorpus.org

erensezener replied:
Removed the special cases 4443294

denisalevi replied:
The special cases increase the speed for ```axis = 0``` and ```axis = 1``` by a factor of approximately 1.5.

Measuring the execution time of ```flip(a, 0)``` using ```timeit``` and for an array ```a = np.arange(100**4).reshape(100,100,100,100)``` I got (from 1000000 loops, best of 3):

```0.865 us``` when using the special cases (returning ```flipud(a)```)
```1.34 us``` when using the current implementation from [abe1735](https://github.com/erensezener/numpy/commit/abe173528514ad71ef01d8d7c0ebf7c92abe1913) using @njsmith suggestion.

Does that still fall into micro-optimization?

shoyer replied:
Yes. 0.5 us is very, very fast.
On Fri, Feb 26, 2016 at 11:32 PM Denis Alevi <notifications@github.com>
wrote:

> In numpy/lib/twodim_base.py
> <https://github.com/numpy/numpy/pull/7346#discussion_r54328158>:
>
> > +
> > +           [[0, 1],
> > +            [2, 3]]])
> > +
> > +    >>> flip(A, 1)
> > +    array([[[2, 3],
> > +            [0, 1]],
> > +
> > +           [[6, 7],
> > +            [4, 5]]])
> > +
> > +    >>> A = np.random.randn(3,4,5)
> > +    >>> np.all(flip(A,2)==A[:,:,::-1,...])
> > +    True
> > +    """
> > +    if axis==0:
>
> The special cases increase the speed for axis = 0 and axis = 1 by a
> factor of approximately 1.5.
>
> Measuring the execution time of flip(a, 0) using timeit and for an array a
> = np.arange(100**4).reshape(100,100,100,100) I got (from 1000000 loops,
> best of 3):
>
> 0.865 us when using the special cases (returning flipud(a))
> 1.34 us when using the current implementation from abe1735
> <https://github.com/erensezener/numpy/commit/abe173528514ad71ef01d8d7c0ebf7c92abe1913>
> using @njsmith <https://github.com/njsmith> suggestion.
>
> Does that still fall into micro-optimization?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/7346/files#r54328158>.
>

madphysicist commented on numpy/lib/twodim_base.py:
```diff
@@ -32,6 +32,76 @@ def _min_int(low, high):
+    else:
+        m = asanyarray(m)
+        if m.ndim < 2:
+            raise ValueError("Input must be >= 2-d.")
```
This error message is a little misleading. If you remove the special cases, you can just check for `m.ndims <= axis`. You should probably be making that check anyway since you probably don't want to allow something like `a.shape = (2, 2); flip(a, 3)`.

madphysicist replied:
Also, since this is a valid branch in the code, add an exception test for it. The example I gave above would work just fine in an `assert_raises`.

erensezener replied:
Changed the exception case here 4443294. I will also add a test.

erensezener replied:
Actually, we have already added the following tests (line 148). They cover the case you suggest, right?

```python
def test_axes(self):
    self.assertRaises(ValueError, flip, ones(4), axis=1)
    self.assertRaises(ValueError, flip, ones((4, 4)), axis=2)
```

madphysicist commented:
Add a note in the release notes. Aside from the possible bug and other comments, nice job, this looks like a really handy addition.

madphysicist commented on numpy/lib/twodim_base.py:
```diff
@@ -32,6 +32,76 @@ def _min_int(low, high):
     return int64
 
 
+def flip(m, axis):
```
Since this is a function that works on arbitrary dimensions, not just 2-d, is this the right file to put it? Perhaps `function_base.py` and `tests/test_function_base.py` would be a better choice?

njsmith commented:
Should `flipud` and `fliplr` be rewritten in terms of this?

mhvk commented:
In principle, this could be made even more general by allowing an iterable for `axis` (i.e., flipping over multiple axes at the same time).  Probably overkill, though.

erensezener commented:
I think I addressed all the issues pointed out besides @madphysicist's last comment:

> Since this is a function that works on arbitrary dimensions, not just 2-d, is this the right file to put it? Perhaps function_base.py and tests/test_function_base.py would be a better choice?

Should I also change the file of the function? What are other opinions?

mhvk commented on numpy/lib/twodim_base.py:
```diff
@@ -32,6 +32,75 @@ def _min_int(low, high):
+    """
+    if not hasattr(m, 'ndim'):
+        m = asarray(m)
+    if m.ndim <= axis:
```
This doesn't cover negative values of `axis`. I would remove this altogether and replace it by making the assignment part of a try/except, as follows:
```
    try:
        indexer[axis] = slice(None, None, -1)
    except IndexError:
        raise ValueError(...)
```
Question: Should this be an `IndexError` rather than a `ValueError`?

shoyer replied:
No, ValueError is definitely the right error to raise for an invalid axis.
On Fri, Feb 26, 2016 at 9:30 PM Marten van Kerkwijk <
notifications@github.com> wrote:

> In numpy/lib/twodim_base.py
> <https://github.com/numpy/numpy/pull/7346#discussion_r54326237>:
>
> > +            [2, 3]]])
> > +
> > +    >>> flip(A, 1)
> > +    array([[[2, 3],
> > +            [0, 1]],
> > +
> > +           [[6, 7],
> > +            [4, 5]]])
> > +
> > +    >>> A = np.random.randn(3,4,5)
> > +    >>> np.all(flip(A,2)==A[:,:,::-1,...])
> > +    True
> > +    """
> > +    if not hasattr(m, 'ndim'):
> > +        m = asarray(m)
> > +    if m.ndim <= axis:
>
> This doesn't cover negative values of axis. I would remove this
> altogether and replace it by making the assignment part of a try/except, as
> follows:
>
>     try:
>         indexer[axis] = slice(None, None, -1)
>     except IndexError:
>         raise ValueError(...)
>
> Question: Should this be an IndexError rather than a ValueError?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/7346/files#r54326237>.
>

madphysicist replied:
Also,  I would recommend adding one more exception test that passes in `axis < -m.ndim`.

denisalevi replied:
added negative axis support with try/except for index error here [bc8ec6b](https://github.com/erensezener/numpy/commit/bc8ec6b72d45666baae9ba96d3074b0fffd63433)

and an exception test here [3dd2a56](https://github.com/erensezener/numpy/commit/3dd2a56d4024bf8192f1d189a01932554ee3c7b9)

mhvk commented on numpy/lib/twodim_base.py:
```diff
@@ -32,6 +32,75 @@ def _min_int(low, high):
+    -----
+    flip(m, 0) is equivalent to flipud(m).
+    flip(m, 1) is equivalent to fliplr(m).
+    flip(m, n) corresponds to ``A[...,::-1,...]`` with ``::-1`` at position n.
```
write "corresponds to `m`" rather than `A`

denisalevi replied:
changed variables here [37c3391](https://github.com/numpy/numpy/commit/37c3391e804a8967e2ed9c152528eeaf1d8ab79c)

erensezener commented:
As far as I can see, @denisalevi handled all the remaining issues. Ready for merge?

shoyer commented on numpy/lib/twodim_base.py:
```diff
@@ -32,6 +32,76 @@ def _min_int(low, high):
+    -------
+    out : array_like
+        A view of `m` with the entries of axis reversed.  Since a view is
+        returned, this operation is :math:`\\mathcal O(1)`.
```
I would simply say "constant time".

erensezener replied:
Done e10be0e

charris commented:
Updated.

charris commented:
Needs mention in `doc/release/1.12.0-notes.rst`.

charris commented:
Also squash the 12 commits into one or two. You can use `git rebase -i HEAD~24` for that followed by a force push `git push --force origin <your branch>`.

shoyer commented on numpy/lib/twodim_base.py:
```diff
@@ -32,6 +32,76 @@ def _min_int(low, high):
+        indexer[axis] = slice(None, None, -1)
+    except IndexError:
+        raise ValueError("axis=%i is invalid for the %i-dimensional input array"
+                         %(axis, m.ndim))
```
per pep8, please add a space after the `%` operator

erensezener commented:
@charris under what title should the release note be? New features?

denisalevi commented:
@charris We moved the files and squashed the commits. If you prefer a single PR for flip and rot90, the rot90 PR ([#7247](https://github.com/numpy/numpy/pull/7347#issuecomment-193497201)) includes the changes of this one here (the generalized_rot90 branch is rebased on the generalized_flip branch)

charris commented on doc/release/1.12.0-notes.rst:
```diff
@@ -99,6 +99,10 @@ keyword argument. It can be set to False when no write operation
 
+Generalized ``flip``
+~~~~~~~~~~~~~~~~~~~~
+``flipud`` and ``fliplr`` reverse the elements of an array along axis=0 and axis=1 respectively. The newly added ``flip`` function reverses the elements of an array along any given axis.
```
Lines less than 80 characters in length, need line breaks here.

charris commented:
Getting close. Add an entry in  `doc/source/reference/routines.array-manipulation.rst` also. Search for `flipud` to see where. The test failures are bogus, so don't worry about them.

denisalevi commented:
@charris done and squashed.

charris commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -23,6 +23,89 @@
+    def test_3d_swap_axis0(self):
+        a = np.array([[[0, 1],
+                       [2, 3]],
+
```
Can leave out the blank lines.

charris commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -23,6 +23,89 @@
+    def test_3d_swap_axis2(self):
+        a = np.array([[[0, 1],
+                      [2, 3]],
+
```
indentation off here and below in `b`.

charris commented:
I've made a few comments in the tests for formatting. I'll put this in and you can do the fixes in the rot90 PR.

charris commented:
Thanks @erensezener and co.

erensezener commented:
Thanks @charris. Did the changes here: 7a57783d84188d24ed18dc16a825b13a556087d3
